src/zenml/integrations/kubernetes/orchestrators/kubernetes_orchestrator_entrypoint.py

```diff
@@ -0,0 +1,221 @@
+"""Entrypoint of the k8s master/orchestrator pod."""
+
+#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+
+import argparse
+import json
+import logging
+import sys
+import threading
+from collections import defaultdict
+from typing import Any, Dict, List
+
+from zenml.integrations.kubernetes.orchestrators import tfx_kube_utils
+from zenml.integrations.kubernetes.orchestrators.utils import (
+    build_base_pod_manifest,
+    update_pod_manifest,
+)
+
+
+def parse_args() -> argparse.Namespace:
+    """Parse entrypoint arguments.
+
+    Returns:
+        argparse.Namespace: Parsed args.
+    """
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--run_name", type=str, required=True)
+    parser.add_argument("--pipeline_name", type=str, required=True)
+    parser.add_argument("--image_name", type=str, required=True)
+    parser.add_argument("--kubernetes_namespace", type=str, required=True)
+    parser.add_argument("--pipeline_config", type=json.loads, required=True)
+    return parser.parse_args()
+
+
+def main() -> None:
+    """Entrypoint of the k8s master/orchestrator pod."""
+    # Log to the container's stdout so it can be streamed by the client.
+    logging.basicConfig(stream=sys.stdout, level=logging.INFO)
+    logging.getLogger().setLevel(logging.INFO)
+    logging.info("Starting orchestration...")
+
+    # Parse / extract args.
+    args = parse_args()
+    pipeline_config = args.pipeline_config
+    step_command = pipeline_config["step_command"]
+    sorted_steps = pipeline_config["sorted_steps"]
+    fixed_step_args = pipeline_config["fixed_step_args"]
+    step_specific_args = pipeline_config["step_specific_args"]
+    step_dependencies = pipeline_config["step_dependencies"]
+
+    # Revert step dependencies so we can easily get subsequent steps of a step.
+    step_successors = defaultdict(list)
+    for step_name, deps in step_dependencies.items():
+        for previous_step_name in deps:
+            step_successors[previous_step_name].append(step_name)
+
+    # Get k8s Core API for running kubectl commands later.
+    core_api = tfx_kube_utils.make_core_v1_api()
+
+    # Build base pod manifest.
+    base_pod_manifest = build_base_pod_manifest(
+        run_name=args.run_name,
+        pipeline_name=args.pipeline_name,
+        image_name=args.image_name,
+    )
+
+    class StepDependencies:
```

Comment:
Why is this a nested class? :-) And we dont have step dependencies yet do we?

Reply:
- `step_dependencies` might be a misleading variable name here, it is a dict {step_name: list_of_previous_step_names} that describes the order of steps in the pipeline DAG.
- It's a nested class to prevent race conditions when modifying it simultaneously in different threads. Right now, for every pipeline step, a new thread is created in the orchestrator pod, which starts the respective step pod, waits for it to finish, and then updates the `step_dependencies` of other steps and starts new threads for all steps that can now be run (after all the previous steps have completed).

Any suggestions how to improve this?

Reply:
Majorly refactored this whole file, should be much clearer now how it is working.